### ISSUE TITLE
keep metrics

### ISSUE THREAD

**Issue by tttp** [cody-greene/node-rabbitmq-client#28]
Hi,

We've been using this great lib for a while to read messages from a queue and call a REST API for each message and last week we had a first "grown up" moment (70k in 12 hours, 11k in 30 min).

Everything worked fine but the default configuration (concurrency 1 and prefetch 2) was a bit too slow and we got a bit of a backlog until we stopped and restarted with concurrency 4 + prefetch 8.

We would like to monitor how many messages were:
- processed (ack)
- requeued (nack+requeued)
- rejected (nack+dead-letter)
- waiting to be processing

the first 3 are simple counters and we could implement it in the client using the library, but as it might be a common need, it might be worth implementing in the library?

for the later it seems there are multiple methods to see how many messages are waiting https://www.rabbitmq.com/queues.html#queue-length, would it be possible to implement the queue.declare-ok + messageCount ? 

and a more open question: in your experience, is this worth dynamically changing the concurrency + prefetch or is there no downside to start directly with the max concurrency+prefetch ?

**Comment by cody-greene** [comment]
I'll consider adding an optional stats collector. 

Queue length:
```javascript  
const rabbit = new Connection()

const ch = await rabbit.acquire()
const {messageCount, consumerCount} = await ch.queueDeclare({
  queue: 'whatever', 
  passive: true
})
await ch.close()
```

The risks of unlimited concurrency are:
- limited resources: At a certain point, you end up doing too much which can lead to each iteration of the event loop getting slower and slower (not enough CPU), or you can run out of memory, or you can use too many database connections, file handles, etc
- worker starvation: In a clustered environment, one worker may claim most of the jobs leaving others with nothing to do.
- out-of-order processing: If messages in your queue should be processed in order then you don't want ANY concurrency.

**Comment by tttp** [comment]
thx. I ended up doing it from the ready event from my consumer

  const sub = rabbit.createConsumer( ...

  const messageCount = async () => {
    const {messageCount} = await sub._ch.queueDeclare({
      queue: sub._queue,
      passive: true
    })
    console.log("messages in the queue",messageCount);
  }

  sub.on('ready', async () => {
   await messageCount();
  });


Note for others/my future self:this number might be slightly off, for instance if you have prefetched messages, so probably better to take is an indication rather than hard truth.

as for the concurrency, indeed, I planned an upper limit, say 10 in parallel, but I was wondering if it was worth the trouble of adjusting dynamically between 1 and 10 or just let it run at 10 all the time, even if there are a few messages.

and good point about out-of-order, it's mostly irrelevant in our case, but we might have some weird edge cases indeed.

X+

**Comment by cody-greene** [comment]
I don't think dynamically changing concurrency is generally worth it. It only makes sense if your workload is highly variable (in terms of memory/cpu) and you want to optimize performance/resource-usage. In which case you could adjust the concurrency value based on current cpu usage & free memory, rather than the number of messages in the queue.

-------- Original Message --------

**Comment by cody-greene** [comment]
Also, I don't recommend accessing undocumented private fields (those with leading underscores). It's like placing a ticking bomb in your application.

-------- Original Message --------

**Comment by tttp** [comment]
> don't recommend accessing undocumented private fields (those with leading underscores)

I found a method to access the [consumer's queue name](https://cody-greene.github.io/node-rabbitmq-client/latest/classes/Consumer.html#queue), is there one to access the channel too?

**Comment by cody-greene** [comment]
Would https://github.com/cody-greene/node-rabbitmq-client/pull/32 address you need for metrics?

Also a few people have requested it so you should have an [easier time with queueDeclare](https://github.com/cody-greene/node-rabbitmq-client/pull/29) now.

**Comment by tttp** [comment]
Neat, thanks!

looping back to the second part of the initial question to fetch how many messages are in the queue:

```
const rabbit = new Connection()

const ch = await rabbit.acquire()
const {messageCount, consumerCount} = await ch.queueDeclare({
  queue: 'whatever', 
  passive: true
})
await ch.close()
```

1. Can we use the lazy channel instead of creating a new one?
2. Could you add a Consumer.waiting and query it when initialising the Consumer?

(rationale: doing it it before starting the Consumer is delaying the "real" work, but trying to do it in parallel means the prefetch are likely to be already done and I didn't find a way to plug the request at the right place in the Consumer initialization)

X+

**Comment by cody-greene** [comment]
> Can we use the lazy channel instead of creating a new one?

Yes

> Could you add a Consumer.waiting and query it when initialising the Consumer?

I'm not convinced that's at all useful. But queueDeclare is always called during setup and I'm just discarding the messageCount so I suppose I could expose it as `consumer.stats.initialMessageCount`

**Comment by cody-greene** [comment]
Included in 4.3.0
